pkg/management/postgres/wal.go

```diff
@@ -97,20 +98,74 @@ func (w *walArchiveBootstrapper) tryBootstrapWal() error {
 	}
 
 	if walArchivingWorking {
-		log.Info("WAL archiving is working, proceeding with the backup")
+		log.Info("WAL archiving is working")
 		return nil
 	}
 
 	if lastFailedTimePresent {
-		log.Info("WAL archiving is not working, will retry in one minute")
+		log.Info("WAL archiving is not working")
 		return errors.New("wal-archive not working")
 	}
 
-	if w.firstWalArchiveTriggered {
-		log.Info("Waiting for the first WAL file to be archived")
-		return errors.New("waiting for first wal-archive")
+	return errNoWalArchivePresent
+}
+
+type walArchiveBootstrapper struct {
+	walArchiveAnalyzer
+	createdFirstWal bool
+}
+
+func newWalArchiveBootstrapperForPrimary() *walArchiveBootstrapper {
+	return &walArchiveBootstrapper{
+		walArchiveAnalyzer: walArchiveAnalyzer{
+			dbFactory: func() (*sql.DB, error) {
+				db, openErr := sql.Open(
+					"pgx",
+					fmt.Sprintf("host=%s port=%v dbname=postgres user=postgres sslmode=disable",
+						GetSocketDir(),
+						GetServerPort(),
+					),
+				)
+				if openErr != nil {
+					log.Error(openErr, "can not open postgres database")
+					return nil, openErr
+				}
+				return db, nil
+			},
+		},
 	}
+}
+
+func (w *walArchiveBootstrapper) ensureFirstWalArchived(backoff wait.Backoff) error {
+	return retry.OnError(backoff, resources.RetryAlways, func() error {
+		db, err := w.dbFactory()
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if closeErr := db.Close(); closeErr != nil {
+				log.Debug("Error while closing connection", "err", closeErr.Error())
+			}
+		}()
+
+		err = w.mustHaveFirstWalArchived(db)
+		if !errors.Is(err, errNoWalArchivePresent) {
+			return err
+		}
+
+		if w.createdFirstWal {
```

Comment:
I think this is very wrong...
And, I don't see where it gets set (other than in test)

Reply:
I think that in the shipWalFile method, we should set `w.createdFirstWal` to true, no?